src/library/autodj/autodjfeature.cpp

```diff
@@ -275,41 +271,61 @@ void AutoDJFeature::slotCrateAutoDjChanged(int crateId, bool added) {
             }
         }
     }
-    // Let subscribers know whether it's possible to add a random track.
-    emit(enableAddRandom(m_crateList.length() > 0));
 #endif // __AUTODJCRATES__
 }
+// Adds a random track : this will be faster when there are sufficiently large 
+// tracks in the crates
 
 void AutoDJFeature::slotAddRandomTrack(bool) {
 #ifdef __AUTODJCRATES__
-    int retriveAttempts = 0;
-    while (retriveAttempts < kMaxRetiveAttempts) {
-        // Get access to the auto-DJ playlist.
-        PlaylistDAO& playlistDao = m_pTrackCollection->getPlaylistDAO();
-        if (m_iAutoDJPlaylistId >= 0) {
+    int failedRetriveAttempts = 0;
+    bool playlistHasTrack = false;
+
+    // Get access to the auto-DJ playlist
+    PlaylistDAO& playlistDao = m_pTrackCollection->getPlaylistDAO();
+    int iTrackId = -1;
+
+    if (m_iAutoDJPlaylistId >= 0) {
+        while (failedRetriveAttempts < kMaxRetriveAttempts) {
             // Get the ID of a randomly-selected track.
-            int iTrackId = m_autoDjCratesDao.getRandomTrackId();
-            //qDebug() << "we got iTrackid " <<iTrackId;
+            iTrackId = m_autoDjCratesDao.getRandomTrackId();
             if (iTrackId != -1) {
                 // Get Track Information
                 TrackPointer addedTrack = (m_pTrackCollection->getTrackDAO()).getTrack(iTrackId);
-                if(addedTrack->exists()) {
-                    //qDebug() << addedTrack->getTitle() << " Exists";
-                    // Add this randomly-selected track to the auto-DJ playlist.
+                playlistHasTrack = playlistDao.playlistHasTrack(iTrackId, m_iAutoDJPlaylistId);
+                if(addedTrack->exists() && !playlistHasTrack) {
```

Comment:
It does actually .
''''
playlistHasTrack = playlistDao.playlistHasTrack(iTrackId, m_iAutoDJPlaylistId);
''''
two lines above is used to check this and incase the track is already queued , the attempt is considered a failure. And another attempt is made. I dont think there is a way to modify the query so that we retrive a fresh track. We have to retrive a track and check it with the playlist_id if it is already added. Ill look again if I missed somethig .

Ill try to pass the already added track_ids as a Qlist in the arguement and modify the query to exclude them , does that sound good?